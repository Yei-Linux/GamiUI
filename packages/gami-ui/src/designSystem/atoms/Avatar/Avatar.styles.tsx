import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { ICustomTheme } from 'providers/ThemeGamification/ThemeGamification'
import { mixinFlexVariants } from 'styles/mixins/flex'
import { InheritGlobalStylesComponent } from 'styles/utilities/commonComponent'
import { validatorProperty } from 'styles/utilities/validatorsCss'

const valideZoomMode = () => css`
  &.zoom__outside,
  &.zoom__inside {
    transition: transform 0.2s;
  }
  &.zoom__outside:hover {
    transform: scale(1.2);
  }
  &.zoom__inside:hover {
    transform: scale(0.8);
  }
`

export const Avatar = InheritGlobalStylesComponent(styled.div<{
  $borderColor?: string
  $background?: string
  $textColor?: string
  theme?: ICustomTheme
}>`
  width: ${({ theme }) => theme.tokens.sizes.avatar.width};
  height: ${({ theme }) => theme.tokens.sizes.avatar.height};

  background: ${({ $background }) => $background};
  overflow: hidden;

  ${({ $textColor }) => validatorProperty('color', $textColor)}
  ${({ $borderColor }) =>
    validatorProperty('border-color', `1px solid ${$borderColor}`)}

  ${valideZoomMode()}

  ${mixinFlexVariants({ justifyContent: 'center', alignItems: 'center' })};
`)
