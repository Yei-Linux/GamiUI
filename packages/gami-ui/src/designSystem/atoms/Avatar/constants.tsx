export const options = {
  type: [] as const,
  zoomMode: ['outside', 'inside', 'none'] as const,
}

export const avatarProfilesImages = {
  one: 'https://i.imgur.com/vyW3w3r.png',
  two: 'https://i.imgur.com/lkIgoZP.png',
  three: 'https://i.imgur.com/fuFwKbQ.png',
  four: 'https://i.imgur.com/IXPZxIr.png',
  five: 'https://i.imgur.com/ub0bTXj.png',
}
