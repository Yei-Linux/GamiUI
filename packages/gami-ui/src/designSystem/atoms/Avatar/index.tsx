export { default } from './Avatar'
export { options } from './constants'
