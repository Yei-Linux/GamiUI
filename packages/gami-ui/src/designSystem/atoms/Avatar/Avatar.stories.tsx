import React from 'react'
import { Meta } from '@storybook/react'

import Avatar from '.'
import Icon from '../Icon'
import { avatarProfilesImages } from './constants'
import { getListTemplateTypes } from 'core/helpers/storybook.helper'
import { DESIGN_TYPES } from 'core/utils/constants'

export default {
  title: 'Atoms/Avatar',
  component: Avatar,
  args: {},
  argTypes: {},
} as Meta

const ListTemplateType = getListTemplateTypes(Avatar)

export const WithImage = ListTemplateType.bind({})
WithImage.args = {
  variants: {
    examples: Object.entries(avatarProfilesImages).map(([, value], index) => ({
      label: `Image ${index + 1}`,
      value,
    })),
    field: 'src',
  },
  zoomMode: 'inside',
}

export const WithBorders = ListTemplateType.bind({})
WithBorders.args = {
  variants: {
    examples: DESIGN_TYPES.border.map((border) => ({
      label: border,
      value: border,
    })),
    field: 'border',
  },
  src: avatarProfilesImages.one,
  zoomMode: 'inside',
}

export const WithShadows = ListTemplateType.bind({})
WithShadows.args = {
  variants: {
    examples: DESIGN_TYPES.shadow.map((shadow) => ({
      label: shadow,
      value: shadow,
    })),
    field: 'shadow',
  },
  src: avatarProfilesImages.one,
  zoomMode: 'inside',
  border: 'lg',
}

export const WithText = ListTemplateType.bind({})
WithText.args = {
  variants: {
    examples: [
      {
        label: 'Textone',
        value: 'Jesus',
        customProps: {
          background: '#54BAB9',
          textColor: 'white',
        },
      },
      {
        label: 'Texttwo',
        value: 'Yei Linux',
        customProps: {
          background: '#54BAB9',
          textColor: 'white',
        },
      },
      {
        label: 'Textthree',
        value: 'Cesar Jesus Alberto',
        customProps: {
          background: '#54BAB9',
          textColor: 'white',
        },
      },
    ],
    field: 'text',
  },
}
export const WithIcon = ListTemplateType.bind({})
WithIcon.args = {
  variants: {
    examples: [
      {
        label: 'Icon 1',
        value: <Icon fill="none" size="25px" name="facebook__transparent" />,
      },
      {
        label: 'Icon 2',
        value: <Icon fill="none" size="25px" name="facebook__transparent" />,
      },
      {
        label: 'Icon 3',
        value: <Icon fill="none" size="25px" name="facebook__transparent" />,
      },
    ],
    field: 'icon',
  },
  background: 'white',
  shadow: 'md',
}
