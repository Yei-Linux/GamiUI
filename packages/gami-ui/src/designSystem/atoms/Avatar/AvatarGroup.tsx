import React from 'react'
import * as S from './Avatar.styles'

export interface IAvatarGroup {
  children: React.ReactNode[]
  count?: number
  counterBackground?: string
}

const AvatarGroup = ({
  children,
  count,
  counterBackground,
}: IAvatarGroup) => {
  return (
    <S.AvatarGroupStyled>
      {children}
      {count && (
        <S.CountStyled $background={counterBackground}>+{count}</S.CountStyled>
      )}
    </S.AvatarGroupStyled>
  )
}

AvatarGroup.displayName = 'AvatarGroup'

export default AvatarGroup
