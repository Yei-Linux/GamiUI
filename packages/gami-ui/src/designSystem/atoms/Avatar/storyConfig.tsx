import React from 'react'
import {
  getInheritGlobalStylesStories,
  getStoryConfigStructure,
} from 'core/helpers/storybook.helper'
import Avatar from '.'
import Icon from '../Icon'
import { avatarProfilesImages, options, withTextExamples } from './constants'
import { IStoryMainConfig } from 'core/domain/interfaces/IStorybook'

const mainConfig: IStoryMainConfig = {
  title: 'Atoms/Avatar',
  component: Avatar,
  args: {},
  argTypes: {
    src: {
      control: 'text',
      description: 'Image source',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.src,
      },
    },
    text: {
      control: 'text',
      description: 'Display text when image is missing',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.text,
      },
    },
    alt: {
      control: 'text',
      description: 'Display altText in avatar',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.alt,
      },
    },
    background: {
      control: 'color',
      description: 'Change avatar background',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.background,
      },
    },
    textColor: {
      control: 'color',
      description: 'Change avatar text color',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.textColor,
      },
    },
    borderColor: {
      control: 'color',
      description: 'Change avatar border color',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.borderColor,
      },
    },
    zoomMode: {
      control: { type: 'select', options: options.zoomMode },
      description: 'Display avatar zoom',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.zoomMode,
      },
    },
    textMode: {
      control: { type: 'select', options: options.textMode },
      description: 'Display avatar text mode',
      table: {
        type: { summary: 'string' },
        defaultValue: Avatar.defaultProps?.textMode,
      },
    },
  },
}

const storiesInheritGlobalStyles = getInheritGlobalStylesStories({
  border: {
    args: {
      background: '#F76E11',
      src: avatarProfilesImages.one,
      zoomMode: 'inside',
    },
  },
  shadow: {
    args: {
      src: avatarProfilesImages.one,
      zoomMode: 'inside',
      border: 'lg',
    },
  },
  width: {
    args: {
      src: avatarProfilesImages.one,
      height: 'auto',
      zoomMode: 'inside',
      border: 'lg',
    },
  },
  height: {
    args: {
      src: avatarProfilesImages.one,
      width: 'sm',
      zoomMode: 'inside',
      border: 'lg',
    },
  },
  size: {
    args: {
      src: avatarProfilesImages.one,
      zoomMode: 'inside',
      border: 'lg',
    },
  },
  fontWeight: null,
  textAlign: null,
  margin: {
    args: {
      background: '#F76E11',
      src: avatarProfilesImages.one,
      zoomMode: 'inside',
      border: 'lg',
    },
    examples: ['0px', '2px', '4px', '6px', '8px', '10px'],
  },
  padding: {
    args: {
      background: '#F76E11',
      src: avatarProfilesImages.one,
      zoomMode: 'inside',
      border: 'lg',
    },
    examples: ['0px', '2px', '4px', '6px', '8px', '10px'],
  },
})

const storiesComponent = [
  {
    storyName: 'WithGroup',
    parent: {
      args: {
        count: 12,
      },
      variants: [
        {
          label: 'WithImage',
          value: Object.entries(avatarProfilesImages).map(
            ([, value], index) => (
              <Avatar zoomMode="outside" src={value} key={index} />
            )
          ),
        },
        {
          label: 'WithText',
          value: withTextExamples.map(({ value, customProps }, index) => (
            <Avatar
              zoomMode="outside"
              text={value}
              key={index}
              {...customProps}
            />
          )),
        },
      ],
    },
  },
  {
    storyName: 'WithImage',
    self: {
      args: {
        zoomMode: 'inside',
      },
      variants: {
        examples: Object.entries(avatarProfilesImages).map(
          ([, value], index) => ({
            label: `Image ${index + 1}`,
            value,
          })
        ),
        field: 'src',
      },
    },
  },
  {
    storyName: 'WithText',
    self: {
      args: {},
      variants: {
        examples: withTextExamples,
        field: 'text',
      },
    },
  },
  {
    storyName: 'WithIcon',
    self: {
      args: {
        background: 'white',
        shadow: 'md',
      },
      variants: {
        examples: [
          {
            label: 'Icon 1',
            value: (
              <Icon fill="none" size="25px" name="facebook__transparent" />
            ),
          },
          {
            label: 'Icon 2',
            value: (
              <Icon fill="none" size="25px" name="facebook__transparent" />
            ),
          },
          {
            label: 'Icon 3',
            value: (
              <Icon fill="none" size="25px" name="facebook__transparent" />
            ),
          },
        ],
        field: 'icon',
      },
    },
  },
  {
    storyName: 'WithAlt',
    self: {
      args: {
        background: '#F76E11',
        src: avatarProfilesImages.one,
        zoomMode: 'inside',
        shadow: 'md',
      },
      variants: {
        examples: [
          {
            label: 'Alt 1',
            value: 'Example Alt Avatar 1',
          },
          {
            label: 'Alt 2',
            value: 'Example Alt Avatar 2',
          },
          {
            label: 'Alt 3',
            value: 'Example Alt Avatar 3',
          },
        ],
        field: 'alt',
      },
    },
  },
  {
    storyName: 'WithSetsAndBreakpoints',
    self: {
      args: {
        breakpoints: [
          { mediaMaxWidth: '320px', size: '280px' },
          { mediaMaxWidth: '480px', size: '440px' },
          { size: '800px' },
        ],
        background: '#F76E11',
        src: avatarProfilesImages.one,
        zoomMode: 'inside',
        shadow: 'md',
      },
      variants: {
        examples: [
          {
            label: 'For Sets device',
            value: [
              { setSrc: avatarProfilesImages.one, setWidth: '320w' },
              { setSrc: avatarProfilesImages.one, setWidth: '480w' },
              { setSrc: avatarProfilesImages.one, setWidth: '800w' },
            ],
          },
        ],
        field: 'sets',
      },
    },
  },
  {
    storyName: 'WithZoomMode',
    self: {
      args: {
        background: '#F76E11',
        src: avatarProfilesImages.one,
      },
      variants: {
        examples: options.zoomMode.map((mode) => ({
          label: mode,
          value: mode,
        })),
        field: 'zoomMode',
      },
    },
  },
  {
    storyName: 'WithTextMode',
    self: {
      args: {
        text: 'Jesus Alvan',
      },
      variants: {
        examples: options.textMode.map((mode) => ({
          label: `${mode}: (Jesus Alvan)`,
          value: mode,
        })),
        field: 'textMode',
      },
    },
  },
]

const storyConfig = getStoryConfigStructure({
  mainConfig,
  storiesInheritGlobalStyles,
  storiesComponent,
  component: Avatar,
  parentComponent: Avatar.Group,
})

export { storyConfig }
