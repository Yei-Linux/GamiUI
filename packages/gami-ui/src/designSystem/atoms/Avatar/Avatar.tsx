import React from 'react'
import { IGeneralProps } from 'core/domain/interfaces/IGeneralProps'
import Image from '../Image'
import * as S from './Avatar.styles'

import { getGenericPropStyles } from 'styles/utilities/genericPropStyles'
import { sizes } from 'styles/tokens/sizes'
import { REGEX_RULES } from 'core/utils/constants'
import withDefaults from 'hocs/WithDefault'

export interface IAvatar extends IGeneralProps {
  /**
   * Image source
   */
  src?: string
  /**
   * Show icon in avatar
   */
  icon?: React.ReactNode
  /**
   * 	Display text when image is missing
   */
  text?: string
  /**
   * 	Display altText in avatar
   */
  alt?: string
  /**
   * Change avatar background
   */
  background?: string
}

const Avatar = ({
  src,
  icon,
  text,
  alt,
  background = 'rgb(97, 104, 106)',
  ...genericsProps
}: IAvatar) => {
  const showOnlyFirstLettersOnText = (text: string): string => {
    const firstLettersCaptured: string[] | null = text.match(
      REGEX_RULES.CATCH_FIRST_LETTERS
    )
    const joinFirstLetters = firstLettersCaptured?.join('') || ''

    return joinFirstLetters
  }

  const renderContent = (): React.ReactNode => {
    if (icon) return icon
    if (text) return showOnlyFirstLettersOnText(text)
    if (src)
      return (
        <Image
          width={sizes.avatar.width}
          height={sizes.avatar.height}
          src={src}
          alt={alt}
        />
      )
    return null
  }

  return (
    <S.Avatar $background={background} {...getGenericPropStyles(genericsProps)}>
      {renderContent()}
    </S.Avatar>
  )
}

const defaultProps = {
  border: 'full',
  fontWeight: 'bold',
  textAlign: 'center',
}

export default withDefaults(Avatar, defaultProps)
