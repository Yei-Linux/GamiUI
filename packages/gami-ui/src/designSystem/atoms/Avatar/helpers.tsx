import { TextModeType } from 'core/domain/types'
import { avatarTextModes } from './constants'
import { REGEX_RULES } from 'core/utils/constants'

export const validateMaxLetters = (
  firstLettersGot: string,
  maxSizeLetters: number
) => {
  const size = firstLettersGot.length

  if (size > maxSizeLetters) {
    return firstLettersGot.substring(0, maxSizeLetters)
  }

  return firstLettersGot
}

export const showOnlyFirstLettersOnText = (
  text: string,
  type: TextModeType,
  maxSizeLetters: number
): string => {
  const regexType = avatarTextModes[type]
  const firstLettersCaptured: string[] | null = text.match(
    REGEX_RULES[regexType]
  )

  const joinFirstLetters = firstLettersCaptured?.join('') || ''

  return validateMaxLetters(joinFirstLetters, maxSizeLetters)
}
