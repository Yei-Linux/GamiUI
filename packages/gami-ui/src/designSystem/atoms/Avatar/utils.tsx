import React from 'react'
import Image from '../Image'
import { showOnlyFirstLettersOnText } from './helpers'
import { TAvatarComponent } from './type'

export type TRenderContent = Pick<
  TAvatarComponent,
  'icon' | 'text' | 'textMode' | 'src' | 'sets' | 'breakpoints' | 'alt'
> & {
  maxSizeLetters: number
}

export const renderContent = ({
  icon,
  text,
  textMode,
  src,
  sets,
  breakpoints,
  alt,
  maxSizeLetters,
}: TRenderContent): React.ReactNode => {
  if (icon) return icon
  if (text && textMode)
    return showOnlyFirstLettersOnText(text, textMode, maxSizeLetters)
  if (src)
    return (
      <Image
        sets={sets}
        breakpoints={breakpoints}
        width="100%"
        height="100%"
        src={src}
        alt={alt}
      />
    )
  return null
}
