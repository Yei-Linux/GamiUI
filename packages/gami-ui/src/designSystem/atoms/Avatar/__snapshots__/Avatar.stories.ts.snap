// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Avatar With Icon 1`] = `
<div
  className="css-psf7un-AvatarWrapper eklgmvl0"
  fontWeight="BOLD"
>
  <i
    className="css-6tw2qr-IconWrapper e1cr9qb30"
  >
    <svg
      fill="white"
      height="25px"
      viewBox="0 0 20 20"
      width="25px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.46 18H11.39V10.7H13.84L14.21 7.86H11.39V6.04C11.39 5.22 11.62 4.66 12.8 4.66H14.31V2.11C14.05 2.08 13.16 2 12.12 2C9.94 2 8.46 3.33 8.46 5.76V7.86H6V10.7H8.46V18Z"
      />
    </svg>
  </i>
</div>
`;

exports[`Storyshots Atoms/Avatar With Src 1`] = `
<div
  className="css-psf7un-AvatarWrapper eklgmvl0"
  fontWeight="BOLD"
>
  <img
    alt="Avatar"
    className="css-1moa472-Img efl6h300"
    height="40px"
    src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRPegjNNBzO3-k5t6hTm1htv1kxyWSOtCRr3A&usqp=CAU"
    width="40px"
  />
</div>
`;

exports[`Storyshots Atoms/Avatar With Text 1`] = `
<div
  className="css-psf7un-AvatarWrapper eklgmvl0"
  fontWeight="BOLD"
  style={
    Object {
      "color": "white",
    }
  }
>
  YL
</div>
`;
